### PR TITLE
[FIX] stock, mrp_subcontracting_purchase: fix subcontract location after destination update

### PR DESCRIPTION
**Pull request by kitan191**
**Steps to reproduce:**
- Install mrp and purchase
- Go to "Inventory / Configuration / Settings"
- Activate "Storage Locations"
- Go to "Inventory / Configuration / Warehouse Management / Operations Types"
- Edit "Receipts" type by activating "Show Detailed Operations"
- Go to "Manufactoring / Configuration / Settings"
- Activate "Subcontracting"
- Create product: (e.g. Product XYZ)
  * Product Type: Storable Product
- Create a BoM for Product XYZ:
  * BoM Type: Subcontracting
  * Subcontractors: [any] (e.g. Azure Interior)
- Create a PO:
  * Vendor: Azure Interior
  * Products: 2 x Product XYZ
- Confirm the PO
- Open the picking from PO via the Receipt smart button
- In "Operations" tab, set done to 1
- On the picking form, change the destination location (e.g. WH/Stock/Shelf1)
- Save
- In "Detailed Operations" tab, a line should have appeared
- Select the same destination location on that line (i.e. WH/Stock/Shelf1)
- Validate the picking and create a backorder for the remaining quantity to produce
- Go to "Inventory / Reporting / Locations"
- Check the locations of Product XYZ (Search Product: XYZ - Group by: Location)
=> The "On Hand Quantity" for Product XYZ is as followed:
     * Virtual Locations/Production: -1.00 (correct)
     * WH/Stock/Shelf1: 1.00 (correct)
- Open the backorder picking from PO via the Receipt smart button
- Record the production of the remaining unit
- Validate the picking
- Go to "Inventory / Reporting / Locations"
- Check the locations of Product XYZ

**Issue:**
The "On Hand Quantity" for Product XYZ is as followed:
* Partners/Vendors: -1.00 (incorrect, it should be empty)
* Physical Locations/Subcontracting Location: 1.00 (incorrect, it should be 0.00)
* Virtual Locations/Production: -2.00 (correct)
* WH/Stock/Shelf1: 2.00 (correct)

**Cause:**
When the PO is confirmed, the stock picking and the stock move are created, they both have the same source and destination locations.
However, in an overridden method from "mrp_subcontracting" module, a check is performed on the move to determine if it is a subcontract.
If it is the case, its source location is set to the subcontractor location and so, the source location of the picking and the move is not the same anymore.
When the destination location is changed on the picking, an onchange is triggering an update of the destination location AND the source location of the move to the values coming from the picking, erasing the subcontractor location set on the move.
The issue only happens for the backorder, because the source location update is not propagated to the stock move lines.
In the case of the original picking, the move lines were already created with the subcontractor location as source location.
But when the backorder is created, the move lines are created with the values coming from a move without the subcontractor location.

**Solution:**
Do not propagate "location_id" from the picking to the subcontracting moves.

opw-3777379

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_picking.py [hunk 1]
```diff
@@ -769,11 +769,28 @@ def _onchange_picking_type(self):
...
+
+    @api.onchange('location_id')
+    def _onchange_location_id(self):
         (self.move_ids | self.move_ids_without_package).update({
             "location_id": self.location_id,
-            "location_dest_id": self.location_dest_id
+        })
+        if any(line.reserved_qty or line.qty_done for line in self.move_ids.move_line_ids):
+            return {'warning': {
+                    'title': 'Locations to update',
```
Comment (nle-odoo) [inline addons/stock/models/stock_picking.py:785]: it was like this before your commit, but a little odd that the title is not translated

**Review by nle-odoo — APPROVED** [review]
looks good to me

it's a little sad that the code is duplicated if it's the same (but adding a third utility  method would not decrease the code too much)

File: addons/stock/models/stock_picking.py (hunk 1 above)
Comment (kitan191) [reply addons/stock/models/stock_picking.py]: Thanks for the review.

I've made it translatable to prevent a potential ticket about it :)

**Comment by kitan191** [comment]
@odoo/rd-logistics

**Comment by Whenrow** [comment]
@kitan191 I would do an override in subcontracting to not update the source location in case of subcontracting move

**Comment by kitan191** [comment]
@Whenrow Thanks for the review.

Do you mean something like this in this file https://github.com/odoo/odoo/blob/16.0/addons/mrp_subcontracting/models/stock_picking.py ?

```python
    @api.onchange('location_id', 'location_dest_id')
    def _onchange_locations(self):
        (self.move_ids.filtered(lambda m: m.is_subcontract) | self.move_ids_without_package.filtered(lambda m: m.is_subcontract)).update({
            "location_dest_id": self.location_dest_id
        })
        (self.move_ids.filtered(lambda m: not m.is_subcontract) | self.move_ids_without_package.filtered(lambda m: not m.is_subcontract)).update({
            "location_id": self.location_id,
            "location_dest_id": self.location_dest_id
        })
        if any(line.reserved_qty or line.qty_done for line in self.move_ids.move_line_ids):
            return {'warning': {
                    'title': _("Locations to update"),
                    'message': _("You might want to update the locations of this transfer's operations")
                }
            }
```

**Comment by Whenrow** [comment]
robodoo r+

**Pull request closed without merging**